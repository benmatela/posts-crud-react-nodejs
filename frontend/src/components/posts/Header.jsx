import React from "react";
import { Link } from "react-router-dom";

export const Header = () => {
  return (
      <div className="w-full text-black">
        <div id="links" data-testid={"links"} className="max-w-[1240px] mt-3 mx-auto grid grid-cols-2 text-center">
          <Link to="/posts" data-testid={"link"} className="hover:underline link">
            LIST POSTS
          </Link>
          <Link to="/posts/add" data-testid={"link"} className="hover:underline link">
            ADD POST
          </Link>
        </div>
      </div>
  );
};
