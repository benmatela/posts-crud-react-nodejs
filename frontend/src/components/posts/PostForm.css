.MuiInputBase-input {
    width: 300px;
}

.MuiInputBase-root {
    margin-top: 10px;
}

textarea.form-input {
    border: 1px rgba(103, 100, 100, 0.54) solid;
    border-radius: 5px;
}