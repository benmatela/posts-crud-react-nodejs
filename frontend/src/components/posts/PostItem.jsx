import React from "react";

export const PostItem = (props) => {
  const handleClick = (event) => {
    const index = event.currentTarget.id;
    props.handlePostItemClick(index);
  };

  return props.posts.map((post, i) => {
    return (
      <p
        id={post.id}
        onClick={(event) => handleClick(event)}
        key={i}
        className={
          "py-1 cursor-pointer hover:bg-[#e3e1d2] my-1 justify-evenly bg-[#F0EABE] text-[#000300] w-[100%] md:grid md:grid-cols-2 post-item rounded-md font-medium"
        }
      >
        <span id={"title" + post.id} className="m-2">
          ({post.id}). {post.title}
        </span>
        <span id={"body" + post.id} className="text-gray-600 m-2 hidden md:flex">
          {String(post.body).slice(0, 82) + "..."}
        </span>
      </p>
    );
  });
};
