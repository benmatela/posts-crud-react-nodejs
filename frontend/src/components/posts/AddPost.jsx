import React, { useState } from "react";
import PostModel from "./models/post.model";
import { PostForm } from "./PostForm";
import { addPost } from "../../services/post";

export const AddPost = () => {
  const [selectedPost] = useState({ selected: new PostModel(0, 0, "", "") });
  const [submitting, setSubmitting] = useState(false);
  const [successMessage, setSuccessMessage] = useState("");
  const [errorMessage, setErrorMessage] = useState("");
  const loaderSubmitMessage = "Adding new item...";

  /**
   * Handle PostForm 'submit' event and return data
   * @param {*} data
   */
  const handleSubmitClick = async (data) => {
    setSubmitting(true);
    setSuccessMessage("");
    setErrorMessage("");
    const newPost = await addPost(data);
    if (newPost.status === 200) {
      setSuccessMessage(`Post Created with ID: ${newPost.data.id}.`);
    } else {
      setErrorMessage("Error while creating a post.");
    }
    setSubmitting(false);
  };

  return (
    <div className="w-full h-[600px]">
      <h1
        data-testid={"pageHeader"}
        className="text-center text-gray-600 text-3xl font-bold"
        id="pageHeader"
      >
        Add Post
      </h1>
      <PostForm
        mode={"add"}
        handleSubmitClick={handleSubmitClick}
        submitting={submitting}
        selectedPost={selectedPost}
        successMessage={successMessage}
        errorMessage={errorMessage}
        loaderSubmitMessage={loaderSubmitMessage}
      />
    </div>
  );
};
