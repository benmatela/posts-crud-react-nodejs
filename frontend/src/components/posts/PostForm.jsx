import React, { useState } from "react";
import { useForm } from "react-hook-form";
import Input from "@material-ui/core/Input";
import InputLabel from "@material-ui/core/InputLabel";
import TextareaAutosize from "@material-ui/core/TextareaAutosize";
import "./PostForm.css";
import PostModel from "./models/post.model";

export const PostForm = (props) => {
  const [isUpdate] = useState(props.mode === "update" ? true : false);
  const [selectedPost, setSelectedPost] = useState(props.selectedPost.selected);
  const {
    register,
    handleSubmit,
    formState: { errors },
  } = useForm();

  const handleSubmitClick = async (event) => {
    event.preventDefault();
    props.handleSubmitClick(selectedPost);
  };

  const handleDeletePostClick = async (event) => {
    // 'Yes/No' confirmation before 'handleDeletePostClick' is fired
    event.preventDefault();
    props.handleDeletePostClick(selectedPost);
  };

  const handleTitleInpuChange = async (event) => {
    const post = new PostModel(
      selectedPost.userId,
      selectedPost.id,
      event.target.value,
      selectedPost.body
    );
    setSelectedPost(post);
  };

  const handleBodyInputChange = async (event) => {
    const post = new PostModel(
      selectedPost.userId,
      selectedPost.id,
      selectedPost.title,
      event.target.value
    );
    setSelectedPost(post);
  };

  return (
    <form
      id={props.mode + "Form"}
      name={props.mode + "Form"}
      className="bg-white h-min mt-10"
    >
      <div className="text-center">
        <br />
        {/*  FORM FIELDS */}
        <InputLabel id="titleLabel" className="m-2">
          Title:
        </InputLabel>
        <Input
          id="title"
          className="form-input w-[100%] md:w-[50%]"
          defaultValue={selectedPost?.title}
          {...register("title", { required: true })}
          onChange={handleTitleInpuChange}
        />
        <br />
        <small className="text-[#d9534f]">
          {errors.title && <span>Title is required</span>}
        </small>
        <InputLabel id="bodyLabel" className="mt-4 mb-2">
          Body:
        </InputLabel>
        <TextareaAutosize
          id="body"
          className="form-input w-[100%] md:w-[50%]"
          defaultValue={selectedPost?.body}
          label={"Body"}
          {...register("body", { required: true })}
          minRows={5}
          onChange={handleBodyInputChange}
        />
        <br />
        <small className="text-[#d9534f]">
          {errors.body && <span>Body is required</span>}
        </small>
        <br />

        {/*  BUTTONS */}
        {props.submitting ? (
          <h1 className="text-center text-[#28a745] font-bold">Saving...</h1>
        ) : (
          <div>
            <button
              id="submitBtn"
              className="hover:bg-[#67ade3] bg-[#3120E0] mb-4 mt-2 h-10 w-80 rounded-md text-white"
              onClick={(event) => handleSubmit(handleSubmitClick(event))}
              disabled={props.deleting}
            >
              Submit
            </button>
            <br />
          </div>
        )}
        {isUpdate ? (
          !props.deleting ? (
            <button
              id="deleteBtn"
              className="hover:bg-[#e08683] bg-[#d9534f] mb-4 mt-2 h-10 w-80 rounded-md text-white"
              onClick={(event) => handleDeletePostClick(event)}
              disabled={props.submitting}
            >
              Delete
            </button>
          ) : (
            <h1 className="text-center text-[#d9534f] font-bold">
              Deleting...
            </h1>
          )
        ) : null}

        {/*  ERROR MESSAGES */}
        {props.successMessage.length > 0 ? (
          <h1 className="text-center text-[#28a745] font-bold">
            {props.successMessage}
          </h1>
        ) : null}
        {props.errorMessage.length > 0 ? (
          <h1 className="text-center text-[#dc3545] font-bold">
            {props.errorMessage}
          </h1>
        ) : null}
      </div>
    </form>
  );
};
