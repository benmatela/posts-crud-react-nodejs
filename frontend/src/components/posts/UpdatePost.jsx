import React, { useState } from "react";
import { useLocation } from "react-router-dom";
import { useNavigate } from "react-router-dom";
import { deletePostById, updatePost } from "../../services/post";
import PostModel from "./models/post.model";
import { PostForm } from "./PostForm";

export const UpdatePost = () => {
  const navigate = useNavigate();
  const { state } = useLocation();
  const [selectedPost] = useState(state || {
    selected: new PostModel(0, 0, "", ""),
  });
  const [submitting, setSubmitting] = useState(false);
  const [deleting, setDeleting] = useState(false);
  const [successMessage, setSuccessMessage] = useState("");
  const [errorMessage, setErrorMessage] = useState("");
  const [loaderSubmitMessage, setLoaderSubmitMessage] = useState("");
  const [loaderDeleteMessage, setLoaderDeleteMessage] = useState("");

  /**
   * Handle PostForm 'submit' event and return data
   * @param {*} data
   */
  const handleSubmitClick = async (data) => {
    setLoaderSubmitMessage("Updating item...");
    setLoaderDeleteMessage("");
    setSubmitting(true);
    setSuccessMessage("");
    setErrorMessage("");
    const updated = await updatePost(data);
    if (updated.status === 200) {
      setSuccessMessage(`Post successfully updated. ID: ${data.id}.`);
    } else {
      setErrorMessage(
        `Error while attempting to update a post. ID: ${data.id}.`
      );
    }
    setLoaderDeleteMessage("");
    setLoaderSubmitMessage("");
    setSubmitting(false);
  };

  /**
   * Handle PostForm 'delete' event and return data
   * @param {*} data
   */
  const handleDeletePostClick = async (data) => {
    setLoaderSubmitMessage("");
    setLoaderDeleteMessage("Deleting item...");
    setDeleting(true);
    setSuccessMessage("");
    setErrorMessage("");
    const deleted = await deletePostById(data.id);
    if (deleted.status === 200) {
      setSuccessMessage(`Post successfully deleted. You will be redirected..`);
      setLoaderDeleteMessage("");
      setLoaderSubmitMessage("");
      setTimeout(() => {
        navigate("/posts");
        setDeleting(false);
      }, 2000);
    } else {
      setErrorMessage(
        `Error while attempting to delete a post. ID: ${data.id}`
      );
      setLoaderDeleteMessage("");
      setLoaderSubmitMessage("");
    }
    setDeleting(false);
  };

  return (
    <div className="w-full h-[600px]">
      <h1
        className="text-center text-gray-600 text-3xl font-bold"
        id="pageHeader"
      >
        Update Post
      </h1>
      <PostForm
        mode={"update"}
        handleSubmitClick={handleSubmitClick}
        handleDeletePostClick={handleDeletePostClick}
        selectedPost={selectedPost}
        submitting={submitting}
        deleting={deleting}
        successMessage={successMessage}
        errorMessage={errorMessage}
        loaderSubmitMessage={loaderSubmitMessage}
        loaderDeleteMessage={loaderDeleteMessage}
      />
    </div>
  );
};
