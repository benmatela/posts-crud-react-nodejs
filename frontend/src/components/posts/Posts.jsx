import React, { useEffect, useState } from "react";
import { useNavigate } from "react-router-dom";
import { getPosts } from "../../services/post";
import { PostItem } from "./PostItem";
import { Search } from "../shared/Search";

export const Posts = () => {
  const navigate = useNavigate();
  const [posts, setPosts] = useState([]);
  const [postsStorage, setPostsStorage] = useState([]);
  const [isLoading, setIsLoading] = useState(true);
  const searchPlaceholder = "Search Posts by title";

  useEffect(() => {
    getPosts()
      .catch((error) => {
        // show error message on screen
        setIsLoading(false);
      })
      .then((res) => {
        setPosts(res.data);
        setPostsStorage(res.data);
        setIsLoading(false);
      });
  }, []);

  /**
   * Handle search input and get the searchItem
   * @param {*} searchItem
   */
  const handleSearchClick = (searchItem) => {
    if (String(searchItem).trim().length > 0) {
      setPosts(postsStorage);
      setPosts((current) =>
        current.filter((post) => {
          return String(post.title)
            .trim()
            .toLowerCase()
            .includes(String(searchItem).toLowerCase());
        })
      );
    } else {
      setPosts(() => {
        return postsStorage;
      });
    }
  };

  /**
   * Handle selected post item click
   * @param {*} index
   */
  const handlePostItemClick = (index) => {
    const selected = posts.find((p) => Number(p.id) === Number(index));
    if (selected) {
      navigate("/posts/update", { state: { selected } });
    }
  };

  return (
    <div className="w-full h-[600px] mb-3">
      <h1
        className="text-center text-3xl text-gray-600 font-bold"
        id="pageHeader"
      >
        {`Posts (${posts.length})`}
      </h1>
      {isLoading ? (
        <h1
          id="loader"
          className="text-center text-[#28a745] md:text-4xl text-2xl mt-4 font-bold"
        >
          Loading posts..
        </h1>
      ) : (
        <div className="m-3">
          <Search
            handleSearchClick={handleSearchClick}
            searchPlaceholder={searchPlaceholder}
          />
          <div className="m-3"></div>
          <div className="overflow-y-scroll h-[700px] md:grid md:grid-cols-4" id="postItems">
            <PostItem handlePostItemClick={handlePostItemClick} posts={posts} />
          </div>
        </div>
      )}
    </div>
  );
};
