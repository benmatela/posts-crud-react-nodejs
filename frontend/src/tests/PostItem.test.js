import { cleanup, render, fireEvent } from "@testing-library/react";
import { PostItem } from "../components/posts/PostItem";

const dummyPosts = [
  {
    userId: 1,
    id: 1,
    title:
      "sunt aut facere repellat provident occaecati excepturi optio reprehenderit",
    body: "quia et suscipit\nsuscipit recusandae consequuntur expedita et cum\nreprehenderit molestiae ut ut quas totam\nnostrum rerum est autem sunt rem eveniet architecto",
  },
  {
    userId: 1,
    id: 2,
    title: "qui est esse",
    body: "est rerum tempore vitae\nsequi sint nihil reprehenderit dolor beatae ea dolores neque\nfugiat blanditiis voluptate porro vel nihil molestiae ut reiciendis\nqui aperiam non debitis possimus qui neque nisi nulla",
  },
];

afterEach(cleanup);

describe("Test PostItem component", () => {
  it("should render", () => {
    const component = render(<PostItem posts={dummyPosts} />);

    expect(component).toMatchSnapshot();

    const title = `(${dummyPosts[0].id}). ${dummyPosts[0].title}`;
    expect(document.getElementById(`title${dummyPosts[0].id}`).innerHTML).toEqual(title);

    const posts = document.getElementsByClassName("post-item");
    expect(posts).toHaveLength(2);
  });

  it("should click post item", () => {
    const handlePostItemClick = jest.fn();

    render(
      <PostItem posts={dummyPosts} handlePostItemClick={handlePostItemClick} />
    );

    const button = document.getElementById("1");
    fireEvent.click(button);

    expect(handlePostItemClick).toBeCalled();
  });
});
