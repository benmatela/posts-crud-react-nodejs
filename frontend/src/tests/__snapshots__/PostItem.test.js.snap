// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test PostItem component should render 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="py-1 cursor-pointer hover:bg-[#e3e1d2] bg-[#F0EABE] text-gray-600 w-[100%] mb-1 post-item rounded-md shadow-lg font-medium"
        id="1"
      >
        <div
          class="px-6 py-4 m-2"
        >
          <div
            class="font-bold text-l"
          >
            <span
              class="m-2"
              id="title1"
            >
              (
              1
              ). 
              sunt aut facere repellat provident occaecati excepturi optio reprehenderit
            </span>
          </div>
          <p
            class="m-2 hidden md:flex"
            id="body1"
          >
            quia et suscipit
suscipit recusandae consequuntur expedita et cum
reprehenderit mo...
          </p>
        </div>
      </div>
      <div
        class="py-1 cursor-pointer hover:bg-[#e3e1d2] bg-[#F0EABE] text-gray-600 w-[100%] mb-1 post-item rounded-md shadow-lg font-medium"
        id="2"
      >
        <div
          class="px-6 py-4 m-2"
        >
          <div
            class="font-bold text-l"
          >
            <span
              class="m-2"
              id="title2"
            >
              (
              2
              ). 
              qui est esse
            </span>
          </div>
          <p
            class="m-2 hidden md:flex"
            id="body2"
          >
            est rerum tempore vitae
sequi sint nihil reprehenderit dolor beatae ea dolores neq...
          </p>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="py-1 cursor-pointer hover:bg-[#e3e1d2] bg-[#F0EABE] text-gray-600 w-[100%] mb-1 post-item rounded-md shadow-lg font-medium"
      id="1"
    >
      <div
        class="px-6 py-4 m-2"
      >
        <div
          class="font-bold text-l"
        >
          <span
            class="m-2"
            id="title1"
          >
            (
            1
            ). 
            sunt aut facere repellat provident occaecati excepturi optio reprehenderit
          </span>
        </div>
        <p
          class="m-2 hidden md:flex"
          id="body1"
        >
          quia et suscipit
suscipit recusandae consequuntur expedita et cum
reprehenderit mo...
        </p>
      </div>
    </div>
    <div
      class="py-1 cursor-pointer hover:bg-[#e3e1d2] bg-[#F0EABE] text-gray-600 w-[100%] mb-1 post-item rounded-md shadow-lg font-medium"
      id="2"
    >
      <div
        class="px-6 py-4 m-2"
      >
        <div
          class="font-bold text-l"
        >
          <span
            class="m-2"
            id="title2"
          >
            (
            2
            ). 
            qui est esse
          </span>
        </div>
        <p
          class="m-2 hidden md:flex"
          id="body2"
        >
          est rerum tempore vitae
sequi sint nihil reprehenderit dolor beatae ea dolores neq...
        </p>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
